### ISSUE TITLE
[Bug] [Metadata Utilities] Game is new conditional action doesn't always work

### ISSUE THREAD

**Issue by Dasmius007** [HerrKnarz/Playnite-Extensions#120]
**Describe the bug**
The new "Game is new" condition for conditional actions doesn't always execute properly. If I import a new game from Steam w/o "Download metadata after importing games" enabled in Settings, then it works properly to add a default category or tag. If however I have auto-download metadata enabled on game import, then it doesn't work at all. You also mentioned in the changelog that it should now work for when games are manually added, but even with auto-download metadata disabled, it doesn't work for me either when using "Steam Game Importer" or "Simple Game Importer" plugins to manually add a game. It does work when using the built-in Menu | Add Game | Manually function however.

Also since you removed the "Add Defaults" functionality, there is now no way to use the feature previously available when adding default tags for the checkbox of "Set only when no tags were added on metadata import".

**To Reproduce**
1. Go to Menu | Add-ons | Extension settings | Generic | Metadata Utilities | Conditional actions
2. Click the plus button to add a new conditional action
3. Configure it like the screenshot below
4. Go to Menu | Settings | General
5. Check or uncheck "Download metadata after importing games" to test it both ways
6. Right-click a game in the list from Steam, and click "Remove", then click "Yes"
7. Click Menu | Update Game Library | Steam, OR...
8. Try adding a game with "Steam Game Importer" or "Simple Game Importer" plugins

**Expected behavior**
In all cases, it is expected that the category or tag would be added by default to the games. I am trying to use this new conditional action feature to replace the functionality of how I was using "Add Defaults", but it no longer works as it used to when importing new games w/ auto-metadata import enabled.

**Screenshots**
![Conditional action](https://github.com/user-attachments/assets/62171331-56c3-498b-ac0c-2d7a00969d67)

**please complete the following information:**
 - Windows version: Windows 10 Pro 22H2
 - Playnite version: 10.34
 - Extension version: 1.3.1

**Additional context**
None

**Comment by HerrKnarz** [comment]
It only works, when you have the automatic metadata update disabled, but not when it's enabled? That's strange. It shouldn't work, when the metadata update got canceled somehow, so the addon won't be triggered, but if it is triggered, I don't see, why it wouldn't work. I'll look into it.

**Comment by HerrKnarz** [comment]
I now tested it with the steps you provided, but wasn't able to reproduce the error except for step 8. When importing a game from a library it always set the category for me, with or without metadata download. It also did, when adding a game manually. It only didn't in theses cases, when I aborted the import at some point, but that is expected and would have been the same with the old method.

Do you have some other conditional actions in place, that could remove the category again? It's also possible, that some addon crashes while metadata gets downloaded, so following addons don't get triggered at all. That would at least explain, why it only worked without metadata download. It would be great, if you could test it with only one metadata source and no other addon except for Metadata Utilities enabled, to see if any other addon interferes with it.

I could however reproduce, that the category isn't set when using Simple Game Import, but unfortunately there's nothing I can do about that. Both Steam Game Importer and Simple Game Importer do not trigger the OnLibraryUpdated or ItemUpdated event, but at least one of them is needed for addons to even recognize, that games were added or changed. It also wouldn't have worked with the old method.

**Comment by Dasmius007** [comment]
"Do you have some other conditional actions in place, that could remove the category again?":
**No, I only had the 1 conditional action.**

Regarding Steam Game Importer and Simple Game Importer not triggering events, I understand. To me the solution would be for Playnite itself to trigger these events (or a new, different one) any time a game is imported, and not rely on the add-ons to do this themselves. But this would require a feature request made to Playnite of course.

Per your request to "test it with only one metadata source and no other addon except for Metadata Utilities enabled, to see if any other addon interferes with it", here is what I did:

1. Installed latest Playnite 10.34 in portable mode
2. Enabled Steam only for libraries, left all default settings (so only imported installed games)
3. Just used the default installed IGDB metadata provider
4. Installed latest Metadata Utilities 1.4.1 add-on
5. Configured the conditional action same as above
6. _Note: "Download metadata after importing games" was enabled by default in Settings_
7. Removed 1 game from library, clicking just "Yes", so it was NOT added to the exclusion list
8. Went to Menu | Update Game Library | Steam

**Result:**
It restored the game I removed, downloaded the metadata for it, and added the default category. So it does work to add a category or tag either way now, however there are 2 things I noticed during this:

- Sometimes I have to restart Playnite in order for any changes I make to the conditional action to become effective, such as changing it from adding a category to adding a tag instead (or vice-a-versa), or sometimes enabling/disabling the conditional action (I don't always have to restart Playnite, sometimes it works)
- When I have it add a Tag, there is no way to merge it with the Tags the metadata downloads during import, it simply replaces them (I assume this could be because the add-on triggers first, and then the metadata auto-download process sees the field as non-empty, and skips it since it only downloads missing metadata by default during auto-import I believe; However I thought this worked before, so maybe your add-on is just replacing them)

_Note: Settings | Metadata | Tags is already enabled by default, but the dropdown is empty, so check Official Store in the dropdown to enable it._

Finally regarding "it's also possible, that some addon crashes while metadata gets downloaded, so following addons don't get triggered at all":

- I did have other add-ons installed, and I think they were all disabled, but perhaps one wasn't when I was testing before (I don't recall)
- Also you updated the extension to 1.4.1, and I was using 1.3.1 before, but I updated before this new testing so it's possible you fixed the issue

**Bottom-line:** It is working as expected now except for the 2 issues I mentioned above about having to restart Playnite sometimes for changes, and no way to merge a Tag w/ other tags on metadata import.

**Comment by HerrKnarz** [comment]
> Regarding Steam Game Importer and Simple Game Importer not triggering events, I understand. To me the solution would be for Playnite itself to trigger these events (or a new, different one) any time a game is imported, and not rely on the add-ons to do this themselves. But this would require a feature request made to Playnite of course.

I asked about that on the Playnite discord a while ago, since one user had the same problem with the old default categories feature. IIRC because of the way Playnite handles game imports currently it's not possible to simply trigger the events for Playnite and AFAIK add-ons can't trigger them either without trying to add metadata to games themselves. But Crow, the developer of Playnite said, that in the upcoming Playnite 11 game imports and the corresponding events will be overhauled and should make it possible for add-ons to recognize any newly added game then.

> Sometimes I have to restart Playnite in order for any changes I make to the conditional action to become effective, such as changing it from adding a category to adding a tag instead (or vice-a-versa), or sometimes enabling/disabling the conditional action (I don't always have to restart Playnite, sometimes it works)

Playnite has an issue, that many metadata fields like features, categories, tags etc. don't reliably trigger a refresh of the currently displayed game, when add-ons change them. It usually can be resolved by switching to another game and back. Maybe that's what you were experiencing here?

> When I have it add a Tag, there is no way to merge it with the Tags the metadata downloads during import, it simply replaces them (I assume this could be because the add-on triggers first, and then the metadata auto-download process sees the field as non-empty, and skips it since it only downloads missing metadata by default during auto-import I believe; However I thought this worked before, so maybe your add-on is just replacing them)

Yeah, that is a bug I'm already aware of. I didn't realize, that the ItemUpdated event I use to trigger my actions already runs once before any metadata was downloaded for a game. So adding tags, features etc, with the "Game is new" or even "Is empty" condition currently results in that field being skipped when downloading metadata.

I'm in the process of reworking, how conditional actions trigger and how the "Game is new" condition is working. You will be able to decide for each conditional action, if it runs before downloading metadata for a new game or afterwards. And the "Game is new" condition will check against the list of games present when Playnite was started instead of a timestamp, so all games added in one playnite session will be considered new until you restart Playnite. There also will be a button to reset the new games when necessary without restarting Playnite. I hope the system will be more reliable and flexible then. I'm planning to have the new version ready on the weekend.

**Comment by Dasmius007** [comment]
I am aware of the issue in Playnite, but no that is not what I am experiencing. It took some testing, as the issue is intermittent, but I think I found a reliable way to reproduce it (with or without auto-download of metadata enabled):

1. Configure one conditional action for default Category like the original screenshot above
2. Remove a game from your library (click Yes only, so it is NOT added to the exclusion list)
3. Click Menu | Update Game Library | Steam
4. Click a different game, then click the game that was just re-imported
5. You should see the default Category added correctly
6. Click Menu | Add-ons | Extension settings | Generic | Metadata Utilities (this step is important)
7. Now click Cancel (also important)
8. Open the Metadata Utilities settings again (do NOT do this by skipping the previous 2 steps, its critical you opened it once, canceled the edit, then opened it again)
9. Edit the same conditional action (not make a new one), change it to add a Tag instead of a Category like so: ![Add Tag](https://github.com/user-attachments/assets/a7ad37e7-b3ed-4f39-b56a-436b6d0626a8)
10. Repeat Steps 2 through 4 (with the same game)
11. It still shows the default Category added, even though it should have added a default Tag

**Note:** The key is opening the Metadata Utilities settings, then canceling, and then opening it again before re-configuring the conditional action (just opening the Add-Ons Settings alone won't do it, you must select Metadata Utilities). If you skip opening it and then canceling before opening it again, it won't cause the bug.

**Comment by HerrKnarz** [comment]
I tried to reproduce the issue exactly like you described it with the newest version and it works like it should for me. I changed several things on how a game is registered as "new" and also if a conditional action now is executed before the initial metadata download for a new game or after. So I hope that's also the reason, why I can't reproduce it anymore. Could verify on your end, if it works now for you, too?

**Comment by Dasmius007** [comment]
I tried with your latest version (1.5.1), and yes, it now works properly for me too. I no longer have to restart Playnite for my changes to become active. However there is still 1 outstanding bug, at least for me, regarding default Tags.

I tested the new feature "Execute only on new games before metadata download", both checked and unchecked. Of course for default Category, it works fine (as there is no category info added during metadata import). However when I add a default Tag and have auto-metadata import enabled in Settings, then this new checkbox doesn't work. Regardless if I have it checked or not, it always replaces all the tags on the game with the default Tag I specified. If the checkbox is checked, I would expect this behavior. If unchecked however, I would expect it to merge my default Tag specified with the tags imported during auto metadata update.

Hopefully I don't need to describe repro steps for you (they are pretty simple), but if you need me to I will. I also could create a video showing the bug, or you could even remote into my system and I can show it to you live.

**Comment by HerrKnarz** [comment]
Sorry for the late reply. I didn't see your answer until now.

I just tried it and it works for me. I added a new conditional action, set the condition to "game is new" and the action to add a tag to the game (one that won't be added by other means). When I have the box "before metadata update" checked, it only adds this one tag and doesn't add the ones during metadata import, when I don't have the box checked, it adds the tag to the ones from the metadata import. 

I just released a new version of the addon (1.5.2) that adds some more debug log messages for conditional actions. Can you try it again with that version and having "Write debug log messages" ticked in the addon settings? If it still happens, please send me the extension.log (from the playnite installation folder) and the config.json of the addon (can be found in the addon-settings under "Installed" -> "Generic" - "Metadata Utilities" and clicking "Data folder"). I hope the new debug message will provide enough information to understand what exactly is happening.

**Comment by Dasmius007** [comment]
No worries - I also apologize for the long delay in testing this again.

I tried with the latest add-on version (1.6.0) and Playnite (10.35). I followed your steps, and when "before metadata download" is unchecked it doesn't work properly for me still. It will just have my 1 custom tag, and not merge it with the other tags during metadata import.

So I enabled debug logging, and attached the files here:
[extensions.log](https://github.com/user-attachments/files/18698824/extensions.log)
[config.json](https://github.com/user-attachments/files/18698825/config.json)

For reference, here is how I have the conditional action configured:
![Image](https://github.com/user-attachments/assets/5b2a3585-d4be-4987-b91a-6ebb4aa64d0d)

Here is a screenshot of the game after importing it, with the conditional action configured as above:
![Image](https://github.com/user-attachments/assets/8ee09411-5815-47bb-bdaf-395d5d6c27b7)

Also for reference, here are some of the tags for the game when I disable the conditional action:
![Image](https://github.com/user-attachments/assets/72bd8dc0-0e57-4a8e-8935-f066f0dfe6f0)

I hope this info helps. Please let me know if I can provide anything further, or test it again in a different way.

**Comment by HerrKnarz** [comment]
I finally got some time to test this issue further. I still can't reproduce it, but looking at your log file it seems MetadataUtilities runs the conditional actions two times before the automatic metadata update. The first time no tag is added, since no action is set to "before metadata update", but it is the second time. Only after that the metadata update runs and doesn't add new tags, since the game already has one.

When I try that on my end, the conditional actions only run once before the metadata update. When the action to add the new tag to a new game is run, the other tags are already in the game.

Do you have any other addons installed, that probably change stuff in a game before the metadata is added and triggers Metadata Utilities that way? That's the only possible explanation I can think of here. Unfortunately the extension log doesn't contain any suspicious entries pointing to other addons, but that doesn't mean there aren't any.
